src/json.c

```diff
@@ -187,66 +214,116 @@ int JSON_StoreVectorInDocField(FieldSpec *fs, RedisJSON arr, struct DocumentFiel
     return REDISMODULE_ERR;
   }
 
-  // The right function will put a value of the right type in the address given, or return REDISMODULE_ERR
-  switch (type) {
-    default:
-    case VecSimType_FLOAT32:
-      getElement = (getJSONElementFunc)JSON_getFloat32;
-      break;
-    // Uncomment when support for more types is added
-    // case VecSimType_FLOAT64:
-    //   getElement = (getJSONElementFunc)JSON_getFloat64;
-    //   break;
-    // case VecSimType_INT32:
-    //   getElement = (getJSONElementFunc)JSON_getInt32;
-    //   break;
-    // case VecSimType_INT64:
-    //   getElement = (getJSONElementFunc)japi->getInt;
-    //   break;
-  }
+  getElement = VecSimGetJSONCallback(type);
 
   if (!(df->strval = rm_malloc(fs->vectorOpts.expBlobSize))) {
     return REDISMODULE_ERR;
   }
   df->strlen = fs->vectorOpts.expBlobSize;
 
-  unsigned char step = VecSimType_sizeof(type);
-  char *offset = df->strval;
   // At this point array length matches blob length
-  for (int i = 0; i < arrLen; ++i) {
-    RedisJSON json = japi->getAt(arr, i);
-    if (getElement(json, offset) != REDISMODULE_OK) {
-      rm_free(df->strval);
-      return REDISMODULE_ERR;
-    }
-    offset += step;
+  if (JSON_StoreVectorAt(arr, arrLen, getElement, df->strval, VecSimType_sizeof(type)) != REDISMODULE_OK) {
+    rm_free(df->strval);
+    return REDISMODULE_ERR;
```

Comment:
don't we need validation here?

Reply:
What validation? `JSON_StoreVectorAt` returns `REDISMODULE_OK` only if the vector is valid